File: 3. Behavioral Design Patterns/Template Method/Go/TemplateMethodTemplate.go
package main

import "fmt"

// Template Method Pattern – Go Template

type AbstractClass interface {
	TemplateMethod()
	StepOne()
	StepTwo()
	Hook()
}

type BaseClass struct {
	impl AbstractClass
}

func (b *BaseClass) TemplateMethod() {
	b.impl.StepOne()
	b.impl.StepTwo()
	b.impl.Hook()
}

type ConcreteClassA struct {
	BaseClass
}

func NewConcreteClassA() *ConcreteClassA {
	c := &ConcreteClassA{}
	c.impl = c
	return c
}

func (c *ConcreteClassA) StepOne() { fmt.Println("A: Step 1") }
func (c *ConcreteClassA) StepTwo() { fmt.Println("A: Step 2") }
func (c *ConcreteClassA) Hook()    { /* optional */ }

type ConcreteClassB struct {
	BaseClass
}

func NewConcreteClassB() *ConcreteClassB {
	c := &ConcreteClassB{}
	c.impl = c
	return c
}

func (c *ConcreteClassB) StepOne() { fmt.Println("B: Step 1") }
func (c *ConcreteClassB) StepTwo() { fmt.Println("B: Step 2") }
func (c *ConcreteClassB) Hook()    { fmt.Println("B: Hook") }

func main() {
	NewConcreteClassA().TemplateMethod()
	NewConcreteClassB().TemplateMethod()
}

File: 3. Behavioral Design Patterns/Template Method/Python/TemplateMethodTemplate.py
# Template Method Pattern – Python Template

from abc import ABC, abstractmethod

class AbstractClass(ABC):
    def template_method(self):
        self.step_one()
        self.step_two()
        self.hook()
    
    @abstractmethod
    def step_one(self):
        pass
    
    @abstractmethod
    def step_two(self):
        pass
    
    def hook(self):
        # Optional hook method
        pass

class ConcreteClassA(AbstractClass):
    def step_one(self):
        print("ConcreteClassA: Step 1")
    
    def step_two(self):
        print("ConcreteClassA: Step 2")

class ConcreteClassB(AbstractClass):
    def step_one(self):
        print("ConcreteClassB: Step 1")
    
    def step_two(self):
        print("ConcreteClassB: Step 2")
    
    def hook(self):
        print("ConcreteClassB: Hook")

def main():
    ConcreteClassA().template_method()
    ConcreteClassB().template_method()

if __name__ == "__main__":
    main()
